File: 17. Rearrange chars no adjacent same.cpp
/*
 Task: Rearrange characters in a string so no two adjacent characters are the same
*/

#include <bits/stdc++.h>
using namespace std;

string rearrange(string s)
{
    unordered_map<char, int> cnt;
    for (char c : s) cnt[c]++;
    priority_queue<pair<int, char>> pq;  // max-heap by frequency
    for (auto& p : cnt) pq.push({p.second, p.first});
    string res;
    pair<int, char> prev = {0, '#'};
    while (!pq.empty())
    {
        auto cur = pq.top();
        pq.pop();
        res.push_back(cur.second);
        if (prev.first > 0)
            pq.push(prev);
        cur.first--;
        prev = cur;
    }
    if (res.size() != s.size())
        return "";  // impossible
    return res;
}

int main()
{
    string s = "aaabc";
    cout << "Rearranged: " << rearrange(s) << '\n';
    return 0;
}

// Time Complexity: O(n log k)
// Space Complexity: O(k)

File: 17. Rearrange chars no adjacent same.py
"""
Task: Rearrange characters in a string so no two adjacent characters are the same
Time Complexity: O(n log k)
Space Complexity: O(k)
"""

import heapq
from collections import Counter

def rearrange(s):
    cnt = Counter(s)
    pq = [(-f, ch) for ch,f in cnt.items()]
    heapq.heapify(pq)
    res = []
    prev = None
    while pq:
        f, ch = heapq.heappop(pq)
        res.append(ch)
        if prev:
            heapq.heappush(pq, prev)
            prev = None
        f += 1
        if f < 0:
            prev = (f, ch)
    res = ''.join(res)
    return res if len(res) == len(s) else ''

print('Rearranged:', rearrange('aaabc'))
